src/tcp.rs

```diff
@@ -0,0 +1,50 @@
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+
+/// Configuration for keepalive probes in a TCP stream.
+///
+/// This config's properties map to TCP keepalive properties in Tokio/Mio:
+/// https://github.com/tokio-rs/mio/blob/e6e403fe2a4fc14dfbc74dbb3ae3a14e3044eb6f/src/net/tcp/socket.rs#L25-L46
+///
+/// # Note
+///
+/// Support for the `interval` and `retries` options has just landed in Mio and they are not
+/// available in Tokio yet: https://github.com/tokio-rs/tokio/issues/3082. Setting these currently
+/// has no effect on a TCP stream in vector. These options are parsed into the options config
+/// nevertheless for future support.
+///
+/// Implementing these options would require upgrading the Tokio runtime accordingly first.
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct TcpKeepaliveConfig {
+    #[serde(default)]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(with = "duration")]
+    pub time: Option<Duration>,
+    #[serde(default)]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(with = "duration")]
+    pub interval: Option<Duration>,
+    pub retries: Option<u32>,
+}
+
+mod duration {
```

Comment:
This release suggests this is not needed?https://github.com/serde-rs/serde/releases/tag/v0.8.1

Reply:
Now noticing also that these config options are lacking the `_secs` suffix we usually have for duration parameters.

Reply:
`Duration` is [serialized/deserialized](https://github.com/serde-rs/serde/blob/2b504099e4e480625e74d3599f5ddab2f42c63de/serde/src/de/impls.rs#L1857-L1860) as:

```rust
struct Duration {
    secs: u64,
    nanos: u32,
 }
```

Considering that second precision is more than enough for the timeout, I thought reading/writing the value as a plain `u64` makes more sense in this case. I chose to deserialize as `Duration` to not lose the semantic information of it being a time value.

> Now noticing also that these config options are lacking the _secs suffix we usually have for duration parameters.

Thanks for the hint, I guess it makes sense to do the same if this is the pattern: https://github.com/timberio/vector/blob/95f767d02f92cb6e320b8dc51371dd02959f9499/src/sinks/util/service.rs#L166-L171

Reply:
Done in b0e5caf.